### PR TITLE
WebSocket transport

### PR DESCRIPTION
**Pull request by vaporos**
Add WebSocket-based WsTransport

Implements a websocket-based transport. This transport may be more
network administrator-friendly in some strict enterprise enviroments,
and may be more firewall-friendly and load-balancer-friendly in specific
situations.

Some known flaws:
   - ~~CTRL-C no longer works w/splinterd with this implementation, and
      errors in tests hang. Presumably, there is a thread we need to
      shutdown in those circumstances.~~ This is not repeatable, and was
      possibly not due to ws-transport.
   - ~~The implementation does not reliably pass the mio readiness test
      which is used by the tcp:// and other transports. This is included
      here, but marked #[ignore].~~ The test was updated to be less fragile.

~~This feature should remain experimental until the above two flaws are
fixed.~~

This has not yet been extensively tested on a running splinter network.

### CODE REVIEW

File: libsplinter/Cargo.toml [hunk 1]
```diff
@@ -63,6 +63,7 @@ transact = { version = "0.2", features = ["sawtooth-compat"] }
 url = "1.7.1"
```
Comment (RyanLassigBanks) [inline libsplinter/Cargo.toml:63]: Typo in commit message enviroments/environments

File: libsplinter/Cargo.toml [hunk 2]
```diff
@@ -63,6 +63,7 @@ transact = { version = "0.2", features = ["sawtooth-compat"] }
...
 ursa = { version = "0.1", optional = true }
 uuid = { version = "0.7", features = ["v4"]}
+websocket = { version = "0.24.0", optional = true }
```
[the remark this answers is not in the document]
Comment (vaporos) [reply libsplinter/Cargo.toml]: Fixed

File: libsplinter/Cargo.toml (hunk 1 above)
Comment (vaporos) [reply libsplinter/Cargo.toml:63]: Fixed

File: libsplinter/src/transport/mod.rs [hunk 3]
```diff
@@ -181,73 +184,198 @@ pub mod tests {
...
+                    .values()
+                    .filter(|value| value.is_readable())
+                    .count();
+                writable_count = readiness_map
+                    .values()
+                    .filter(|value| value.is_writable())
+                    .count();
+
+                // We expect to see each connection become both readable and writable. If that
+                // happens, the test is successfull, break out of the loop.
```
Comment (jsmitchell) [inline libsplinter/src/transport/mod.rs:329]: s/successfull/successful/

File: libsplinter/src/transport/mod.rs [hunk 4]
```diff
@@ -181,73 +184,198 @@ pub mod tests {
...
-            connections.push(conn);
+            connections.push(listener.accept().unwrap());
+        }
+
+        // Block until the connector thread tells this thread that polling has been setup. This is
+        // necessary to ensure that we do not lose any readable events.
+        to_listener_rx.recv().unwrap();
+
+        // Send a message on all connections. This should make the poller generate a readable event
+        // with for all the connections.
```
Comment (jsmitchell) [inline libsplinter/src/transport/mod.rs:363]: s/with//

File: libsplinter/src/transport/mod.rs [hunk 5]
```diff
@@ -181,73 +184,198 @@ pub mod tests {
...
+            block!(conn.send(b"hello"), SendError).unwrap();
         }
 
-        // Signal done sending to background thread
-        ready_tx.send(()).unwrap();
+        // Signal the connector thread that this thread has finished sending data on all
+        // connections.
+        to_connector_tx.send(()).unwrap();
 
+        // For each connection, make sure we received the message sent to use from the connector
```
Comment (jsmitchell) [inline libsplinter/src/transport/mod.rs:372]: s/to use//

File: libsplinter/src/transport/mod.rs [hunk 6]
```diff
@@ -181,73 +184,198 @@ pub mod tests {
...
+    /// The process used is essentially:
+    ///
+    /// 1. Create a listener.
+    /// 2. Create a "connector" thread. The original test thread is now the "listener" thread.
+    /// 3. In the connector thread, create some connections, setup polling, and then notify the
+    ///    listener thread.
+    /// 4. In the listener thread, send some data across each connection, then notify the connector
+    ///    thread.
+    /// 5. In the connector thread, loop around poll collecting ready events and determining if we
+    ///    have seen both read and write for each connection. If we have, exist the loop.
```
[the remark this answers is not in the document]
Comment (vaporos) [reply libsplinter/src/transport/mod.rs]: fixed

File: libsplinter/src/transport/mod.rs (hunk 3 above)
Comment (vaporos) [reply libsplinter/src/transport/mod.rs]: fixed

File: libsplinter/src/transport/mod.rs (hunk 4 above)
Comment (vaporos) [reply libsplinter/src/transport/mod.rs]: fixed

File: libsplinter/src/transport/mod.rs (hunk 5 above)
Comment (vaporos) [reply libsplinter/src/transport/mod.rs]: fixed

File: libsplinter/src/transport/mod.rs [hunk 7]
```diff
@@ -132,9 +132,10 @@ pub mod tests {
     use super::*;
     use std::fmt::Debug;
 
+    use std::collections::HashMap;
```
Comment (peterschwarz) [inline libsplinter/src/transport/mod.rs:137]: typo in commit message s/staes/states

Comment (vaporos) [reply libsplinter/src/transport/mod.rs:137]: fixed

**Pull request merged**